src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt

```diff
@@ -150,7 +157,7 @@ internal object LincheckJavaAgent {
             }
 
             // In the model checking mode, Lincheck processes classes lazily, only when they are used.
-            instrumentationMode == MODEL_CHECKING -> {
+            instrumentationMode == MODEL_CHECKING || instrumentationMode == TRACE_RECORDING -> {
```

Comment:
Let's extract this check into `InstrumentationMode.shouldUseLazyTransformation()` (or something like this).